In [ ]:
#INSTALL NECESSARY LIBRERIES
!pip install pandas
!pip install numpy
!pip install -U scikit-learn
!pip install matplotlib
!pip install seaborn

     |████████████████████████████████| 6.9MB 2.9MB/s 
  Found existing installation: scikit-learn 0.22.2.post1
    Uninstalling scikit-learn-0.22.2.post1:
      Successfully uninstalled scikit-learn-0.22.2.post1


### Purpose
The goal is to predict the severity of an accident. We have four values for this variable. Severity is a variable of interest, first of all, for health issues, specifically rapid dispatch of ambulances, then for road prevention issues, since it helps us to understand what we must improve, and finally for the calculation of response for insurance companies.

In [ ]:
#IMPORT LIBRARIES
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [ ]:
#MOUNT DRIVE
from google.colab import drive
drive.mount('/content/drive/')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive/


In [ ]:
#LOAD DATA
us_accident_data = pd.read_csv('/content/drive/My Drive/Accidente.csv')

In [ ]:
#DATASET PREVIEW
us_accident_data.head(n=10)

ID    Source    TMC  ...  Civil_Twilight Nautical_Twilight Astronomical_Twilight
0   A-1  MapQuest  201.0  ...           Night             Night                 Night
1   A-2  MapQuest  201.0  ...           Night             Night                   Day
2   A-3  MapQuest  201.0  ...           Night               Day                   Day
3   A-4  MapQuest  201.0  ...             Day               Day                   Day
4   A-5  MapQuest  201.0  ...             Day               Day                   Day
5   A-6  MapQuest  201.0  ...             Day               Day                   Day
6   A-7  MapQuest  201.0  ...             Day               Day                   Day
7   A-8  MapQuest  201.0  ...             Day               Day                   Day
8   A-9  MapQuest  201.0  ...             Day               Day                   Day
9  A-10  MapQuest  201.0  ...             Day               Day                   Day

[10 rows x 49 columns]

In [ ]:
#AMOUNT OF SAMPLES  
us_accident_data.shape

(2974335, 50)

### Exploratory data analysis

In [ ]:
# DATA STATISTICS
pd.set_option('display.float_format', lambda x: '%.3f' % x) #supress scientific us_accident_data.describe()
us_accident_data.describe()

TMC    Severity  ...  Wind_Speed(mph)  Precipitation(in)
count 2246264.000 2974335.000  ...      2533495.000         975977.000
mean      207.832       2.360  ...            8.298              0.020
std        20.330       0.541  ...            5.139              0.236
min       200.000       1.000  ...            0.000              0.000
25%       201.000       2.000  ...            4.600              0.000
50%       201.000       2.000  ...            7.000              0.000
75%       201.000       3.000  ...           10.400              0.000
max       406.000       4.000  ...          822.800             25.000

[8 rows x 15 columns]

In [ ]:
# PERCENTAGE OF NANs BY ATTRIBUTE
round((100*us_accident_data.isnull().sum()/len(us_accident_data)),2).sort_values().reset_index()

index      0
0                      ID  0.000
1                 Amenity  0.000
2                    Bump  0.000
3                Crossing  0.000
4                Give_Way  0.000
5                Junction  0.000
6                 No_Exit  0.000
7                 Railway  0.000
8              Roundabout  0.000
9                 Station  0.000
10                   Stop  0.000
11        Traffic_Calming  0.000
12         Traffic_Signal  0.000
13           Turning_Loop  0.000
14         Sunrise_Sunset  0.000
15         Civil_Twilight  0.000
16      Nautical_Twilight  0.000
17                Country  0.000
18  Astronomical_Twilight  0.000
19                  State  0.000
20                 Source  0.000
21               Severity  0.000
22             Start_Time  0.000
23               End_Time  0.000
24              Start_Lat  0.000
25              Start_Lng  0.000
26           Distance(mi)  0.000
27            Description  0.000
28                 Street  0.000
29                   City  0.000
30                   Side  0.000
31                 County  0.000
32                Zipcode  0.030
33               Timezone  0.110
34           Airport_Code  0.190
35      Weather_Timestamp  1.230
36         Wind_Direction  1.520
37           Pressure(in)  1.620
38         Temperature(F)  1.880
39            Humidity(%)  1.990
40         Visibility(mi)  2.210
41      Weather_Condition  2.220
42        Wind_Speed(mph) 14.820
43                    TMC 24.480
44          Wind_Chill(F) 62.290
45                 Number 64.470
46      Precipitation(in) 67.190
47                End_Lng 75.520
48                End_Lat 75.520

In [ ]:
#ABSOLUTE NANs
nans_values = us_accident_data.isnull().sum().sort_values().reset_index()
nans_values.columns=['Parameters','Amount of NANS']
nans_values

Parameters  Amount of NANS
0                      ID               0
1                Crossing               0
2                Give_Way               0
3                Junction               0
4                 No_Exit               0
5                 Railway               0
6              Roundabout               0
7                 Station               0
8                    Stop               0
9                 Country               0
10        Traffic_Calming               0
11                  State               0
12                 County               0
13                   Bump               0
14                   Side               0
15         Traffic_Signal               0
16           Distance(mi)               0
17           Turning_Loop               0
18              Start_Lng               0
19              Start_Lat               0
20               End_Time               0
21             Start_Time               0
22               Severity               0
23                 Source               0
24                 Street               0
25                Amenity               0
26            Description               1
27                   City              83
28         Sunrise_Sunset              93
29         Civil_Twilight              93
30  Astronomical_Twilight              93
31      Nautical_Twilight              93
32                Zipcode             880
33               Timezone            3163
34           Airport_Code            5691
35      Weather_Timestamp           36705
36         Wind_Direction           45101
37           Pressure(in)           48142
38         Temperature(F)           56063
39            Humidity(%)           59173
40         Visibility(mi)           65691
41      Weather_Condition           65932
42        Wind_Speed(mph)          440840
43                    TMC          728071
44          Wind_Chill(F)         1852623
45                 Number         1917605
46      Precipitation(in)         1998358
47                End_Lng         2246264
48                End_Lat         2246264

#### Information gathered from _lat, long, postal code, county, state, and city_ 

When we made an analysis of the location data provided, it could be observed that 2126 cases were not correct, that is, one or more of them was not congruent with the associations of the others, that is, county, city, state, or zip code. That is why it was decided to carry out a check with the USPS API, United Post Service, to verify how reliable and complete the data was, based on the latitude and longitude entered. **Why USPS?**  Because it was who registered the postal code in the USA in 1963, and who carried out the postal code expansions.  After forming our query structure with USPS, we will proceed to query the BING API, and we will take into account the points near a neighborhood (square). If any of them at a small distance correspond with the given zip code, we take it as the case of a possible match.To analyze the possible matches of each entry, we give priority to what is entered in the dataset, verifying with our list of probable candidates. First, we start with zip code, and then we go on to analyze state, county, and city. We used the same procedure for the 880 Zipcode and 83 City NANs.

In [ ]:
# FORMAT ZIPCODE.WE DO THIS, BECAUSE MOST OF THE DATASET HAD THE NO COMPLETE FORM,ONLY 4 DIGITS INSTEAD OF 8.
us_accident_data = us_accident_data.assign(Zipcode=us_accident_data.Zipcode.str.split('-').str[0])

In [ ]:
import requests
from time import sleep
#WE BELIEVE THAT USPS IT'S THE MOST RELIABLE SOURCE TO VERITY THE INTEGRITY OF THE DATA
#USPS BRINGS US CITY, STATE, ZIPCODE,BUT NOT COUNTY
def usps_zipcode_api(zipcode:str):
  headers = {
  'authority': 'm.usps.com',
  'pragma': 'no-cache',
  'cache-control': 'no-cache',
  'accept': 'application/json, text/javascript, */*; q=0.01',
  'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36',
  'x-requested-with': 'XMLHttpRequest',
  'sec-fetch-site': 'same-origin',
  'sec-fetch-mode': 'cors',
  'sec-fetch-dest': 'empty',
  'referer': 'https://m.usps.com/m/ZipLookupAction?search=zip',
  'accept-language': 'en,es-419;q=0.9,es;q=0.8',
  'cookie': 'JSESSIONID=0000SL-cA90BnSOvHmSAQ_cW-ZW:1bpccqlcc; NSC_psjhjo-n_443=ffffffff3b2236d945525d5f4f58455e445a4a4212d3',
  }
  params = (('mode', '2'),('tZip', zipcode),('jsonInd', 'Y'),)
  response = requests.get('https://m.usps.com/m/QuickZipAction', headers=headers, params=params)
  location = {}
  for address in response.json()['addresses']:
    location.setdefault(address['state'], set()),
    location[address['state']].add(address['city'])
  for k in location.keys():
    location[k] = sorted(location[k])
  return location

In [ ]:
# LOAD GENERATED ZIPCODES
# WE CREATE TWO DICTS, CITY ---> ZIPCODES, SINCE ONE CITY CAN HAVE MANY ZIPCODES
# AND THE OTHER DICT ZIPCODE ---> CITY
city_to_zipcode = eval(open('/content/drive/My Drive/city_to_zipcode.txt').read())
zipcode_to_city = eval(open('/content/drive/My Drive/zipcode_to_city.txt').read())

In [ ]:
for zipcode in sorted([k for k in us_accident_data.Zipcode.unique() if k is not np.nan]):
  if zipcode not in zipcode_to_city.keys() or len(zipcode_to_city[zipcode].keys()) == 0:
    zipcode_to_city.setdefault(zipcode, {})
    location = usps_zipcode_api(str(zipcode))
    print('missing', zipcode, location)
    for state in location.keys():
      zipcode_to_city[zipcode].setdefault(state, set())
      for city in location[state]:
        zipcode_to_city[zipcode][state].add(city)
        city_to_zipcode.setdefault(state, {})
        city_to_zipcode[state].setdefault(city, set())
        city_to_zipcode[state][city].add(zipcode)
  print(zipcode, zipcode_to_city[zipcode])

Streaming output truncated to the last 5000 lines.
75432 {'TX': {'COOPER'}}
75433 {'TX': {'CUMBY'}}
75440 {'TX': {'EMORY'}}
75442 {'TX': {'FARMERSVILLE'}}
75452 {'TX': {'LEONARD'}}
75453 {'TX': {'LONE OAK'}}
75454 {'TX': {'MELISSA'}}
75455 {'TX': {'MOUNT PLEASANT', 'MT PLEASANT'}}
75457 {'TX': {'MT VERNON', 'MOUNT VERNON'}}
75459 {'TX': {'DORCHESTER', 'HOWE'}}
75460 {'TX': {'PARIS'}}
75462 {'TX': {'RENO', 'PARIS'}}
75469 {'TX': {'PECAN GAP'}}
75470 {'TX': {'PETTY'}}
75473 {'TX': {'POWDERLY'}}
75474 {'TX': {'QUINLAN', 'WEST TAWAKONI'}}
75478 {'TX': {'SALTILLO'}}
75482 {'TX': {'SULPHUR SPGS', 'SULPHUR SPRINGS'}}
75491 {'TX': {'TOM BEAN', 'WHITEWRIGHT'}}
75493 {'TX': {'WINFIELD'}}
75494 {'TX': {'WINNSBORO'}}
75495 {'TX': {'VAN ALSTYNE'}}
75501 {'TX': {'SOUTH TEXARKANA', 'TEXARKANA', 'RED RIVER ARMY DEPOT', 'WAKE VILLAGE', 'S TEXARKANA'}}
75503 {'TX': {'TEXARKANA'}}
75551 {'TX': {'OFARRELL', 'ATLANTA', 'SMYRNA', 'GALLOWAY'}}
75558 {'TX': {'COOKVILLE', 'ARGO'}}
75559 {'TX': {'BEAVER DAM', '

In [ ]:
# SAVE THE FILES FOR FUTURE COMPARISON
with open('/content/drive/My Drive/IA-TP1/zipcode_to_city.txt', 'w') as f:
	f.write(str(zipcode_to_city))
with open('/content/drive/My Drive/IA-TP1/city_to_zipcode.txt', 'w') as f:
	f.write(str(city_to_zipcode))

In [ ]:
# MAKE THE DATASET MATCH USPS FORMAT
us_accident_data = us_accident_data.assign(City=us_accident_data.City.str.upper())
us_accident_data = us_accident_data.assign(County=us_accident_data.County.str.upper())
us_accident_data = us_accident_data.assign(State=us_accident_data.State.str.upper())

In [ ]:
# FIND ERROR CASES
city_mismatch = 0
state_mismatch = 0
mismatch = []
# GENERATE NEW DATAFRAMES WITH COUNTY
zipcode_data = {}
city_data = {}
for zipcode, city, county, state, lat, lng in zip (us_accident_data.Zipcode, us_accident_data.City, us_accident_data.County, us_accident_data.State, us_accident_data.Start_Lat, us_accident_data.Start_Lng):
  if zipcode in zipcode_to_city.keys() and state in zipcode_to_city[zipcode].keys() and city in zipcode_to_city[zipcode][state] and county is not np.nan:
    zipcode_data.setdefault(zipcode,{})
    zipcode_data[zipcode].setdefault(state,{})
    zipcode_data[zipcode][state].setdefault(county,set())
    zipcode_data[zipcode][state][county].add(city)
    city_data.setdefault(state,{})
    city_data[state].setdefault(county,{})
    city_data[state][county].setdefault(city, set())
    city_data[state][county][city].add(zipcode)
  else:
    mismatch.append((zipcode, city, county, state, lat, lng))

    
print('mismatches:', len(mismatch))


mismatches: 2126


In [ ]:
mismatch[:100]

[(nan, 'SOUTHEAST MARIN', 'MARIN', 'CA', 37.885376, -122.516586),
 ('95616', 'DIXON', 'SOLANO', 'CA', 38.536564, -121.74000500000001),
 ('95616', 'DIXON', 'SOLANO', 'CA', 38.536564, -121.74000500000001),
 (nan, 'SOUTHEAST MARIN', 'MARIN', 'CA', 37.885376, -122.516586),
 (nan, 'SOUTHEAST MARIN', 'MARIN', 'CA', 37.885376, -122.516586),
 (nan, 'SOUTHEAST MARIN', 'MARIN', 'CA', 37.885376, -122.516586),
 (nan, 'SAN MATEO', 'SAN MATEO', 'CA', 37.505482, -122.11930100000001),
 (nan, 'SOUTHEAST MARIN', 'MARIN', 'CA', 37.885376, -122.516586),
 ('95616', 'DIXON', 'SOLANO', 'CA', 38.536564, -121.74000500000001),
 (nan, 'SOUTHEAST MARIN', 'MARIN', 'CA', 37.885376, -122.516586),
 ('94525',
  'WEST CONTRA COSTA',
  'CONTRA COSTA',
  'CA',
  38.056346999999995,
  -122.224754),
 ('92372', nan, 'LOS ANGELES', 'CA', 34.451862, -117.660103),
 ('92372', nan, 'LOS ANGELES', 'CA', 34.451862, -117.660103),
 (nan, 'CONSERVATION', 'BROWARD', 'FL', 26.170547, -80.8563),
 (nan, 'CONSERVATION', 'BROWARD', 'FL', 2

In [ ]:
# BING API TO SOLVE CONFLICTED ISSUES
import requests, json

def bing_data(lat, lng, use_locality=False):
	headers = {
    'authority': 'dev.virtualearth.net',
    'pragma': 'no-cache',
    'cache-control': 'no-cache',
    'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36',
    'dnt': '1',
    'accept': '*/*',
    'sec-fetch-site': 'cross-site',
    'sec-fetch-mode': 'no-cors',
    'sec-fetch-dest': 'script',
    'referer': 'https://dev.virtualearth.net/',
    'accept-language': 'en-US,en;q=0.9',
	}

	params = (
    ('o', 'json'),
    ('jsonp', 'Microsoft.Maps.NetworkCallbacks.f574e6'),
    ('key', 'AoltDrgPXkLphU6xf7mwndXly_XEVrnForv0FrqgPHHXwIf-oaB3bfErNvlSYVXc'),
    ('includeEntityTypes', ''),
    ('inclnb', '0'),
    ('incl', ''),
    ('ur', 'ar'),
    ('c', 'en-US'),
	)

	response = requests.get(f'https://dev.virtualearth.net/REST/v1/Locations/{lat},{lng}', headers=headers, params=params)

	data = json.loads(response.text[response.text.find('{'):response.text.rfind('}') + 1])
	state = data['resourceSets'][0]['resources'][0]['address']['adminDistrict']
	county = data['resourceSets'][0]['resources'][0]['address']['adminDistrict2'][:-4].upper()
	zipcode = data['resourceSets'][0]['resources'][0]['address']['postalCode']
	if use_locality:
		locality = data['resourceSets'][0]['resources'][0]['address']['locality'] # WE GENERALLY DONT USE IT BECAUSE USPS HAS A CLEARER DEFINITION OF CITY
		return (state, county, locality, zipcode)
	return (state, county, zipcode)

In [ ]:
from time import sleep
posible_matches = []
index = 0
for m in mismatch:
	lat, lng = m[-2:]
	zcode = m[0]
	p_matches = {}
	print(f'Try with square for index: {index}')
	for d_lat, d_lng in ((0,0),(1e-3,0),(0,1e-3),(-1e-3,0),(0,-1e-3)): #WE CREATE A DIAGONAL, LIKE A NEIGHTBOURHOOD OF POSSIBLE MATCHES
		try:
			case = bing_data(lat + d_lat , lng + d_lng)
			p_matches.setdefault(case, 0)
			p_matches[case] += 1
			if case[-1] == zcode: #IF WE FIND A CASE WHERE THE ZIPCODE FITS WITH THE ZIP GIVEN, WE STOP AND WE ADD IT AS A POSSIBLE AMTCH
				break
		except Exception as e:
			print('Fail', (lat + d_lat , lng + d_lng), 'at index', index)
	index += 1
	posible_matches.append(p_matches)
print(posible_matches)


Try with square for index: 0
Fail (37.885376, -122.516586) at index 0
Fail (37.885376, -122.515586) at index 0
Fail (37.884376, -122.516586) at index 0
Fail (37.885376, -122.51758600000001) at index 0
Try with square for index: 1
Try with square for index: 2
Try with square for index: 3
Fail (37.885376, -122.516586) at index 3
Fail (37.885376, -122.515586) at index 3
Fail (37.884376, -122.516586) at index 3
Fail (37.885376, -122.51758600000001) at index 3
Try with square for index: 4
Fail (37.885376, -122.516586) at index 4
Fail (37.885376, -122.515586) at index 4
Fail (37.884376, -122.516586) at index 4
Fail (37.885376, -122.51758600000001) at index 4
Try with square for index: 5
Fail (37.885376, -122.516586) at index 5
Fail (37.885376, -122.515586) at index 5
Fail (37.884376, -122.516586) at index 5
Fail (37.885376, -122.51758600000001) at index 5
Try with square for index: 6
Fail (37.505482, -122.11930100000001) at index 6
Fail (37.506482, -122.11930100000001) at index 6
Fail (37.50

In [ ]:
# CASES MISSING/WRONG ZIPCODE RETRY WITH GREATER LAT/LNG DISTANCE
index = 0
# TEST DIAGONALS, AND EXPANDED SQUARE/DIAGONALS
area = [((1e-3,1e-3),(-1e-3,1e-3),(1e-3,-1e-3),(-1e-3,-1e-3)), ((1e-2,0),(0,1e-2),(-1e-2,0),(0,-1e-2)), ((1e-2,1e-2),(-1e-2,1e-2),(1e-2,-1e-2),(-1e-2,-1e-2))]
for m, pm in zip(mismatch, posible_matches):
  lat, lng = m[-2:]
  zcode = m[-2:]
  for points in area:
    if zcode not in pm.keys():
      print('TRY MISSING', d_lat , d_lng, 'at index', index)
      p_matches = dict(pm)
      for d_lat, d_lng in points:
        try:
          case = bing_data(lat + d_lat , lng + d_lng)
          p_matches.setdefault(case, 0)
          p_matches[case] += 1
          if case[-1] == zcode:
            break
        except Exception as e:
          print('Fail', *(d_lat , d_lng), 'at index', index)
      posible_matches[index] = p_matches
      pm = p_matches
  index += 1

Streaming output truncated to the last 5000 lines.
TRY MISSING -0.01 -0.01 at index 1559
TRY MISSING -0.001 -0.001 at index 1559
TRY MISSING 0 -0.01 at index 1559
TRY MISSING -0.01 -0.01 at index 1560
TRY MISSING -0.001 -0.001 at index 1560
TRY MISSING 0 -0.01 at index 1560
TRY MISSING -0.01 -0.01 at index 1561
Fail 0.001 0.001 at index 1561
Fail -0.001 0.001 at index 1561
Fail 0.001 -0.001 at index 1561
Fail -0.001 -0.001 at index 1561
TRY MISSING -0.001 -0.001 at index 1561
Fail 0.01 0 at index 1561
Fail 0 0.01 at index 1561
Fail -0.01 0 at index 1561
Fail 0 -0.01 at index 1561
TRY MISSING 0 -0.01 at index 1561
Fail 0.01 0.01 at index 1561
Fail -0.01 0.01 at index 1561
Fail 0.01 -0.01 at index 1561
Fail -0.01 -0.01 at index 1561
TRY MISSING -0.01 -0.01 at index 1562
Fail 0.001 0.001 at index 1562
Fail -0.001 0.001 at index 1562
Fail 0.001 -0.001 at index 1562
Fail -0.001 -0.001 at index 1562
TRY MISSING -0.001 -0.001 at index 1562
Fail 0.01 0 at index 1562
Fail 0 0.01 at index 1562
F

In [ ]:
# CASES MISSING/WRONG ZIPCODE RETRY WITH GREATER LAT/LNG DISTANCE
index = 0
# MISSING CASES TEST DIAGONALS, AND EXPANDED SQUARE/DIAGONALS
area = [((5e-3,0),(0,5e-3),(-5e-3,0),(0,-5e-3)), ((5e-3,2e-3),(-5e-3,2e-3),(5e-3,-2e-3),(-5e-3,-2e-3)),((5e-2,0),(0,5e-2),(-5e-2,0),(0,-5e-2)), ((5e-2,5e-2),(-5e-2,5e-2),(5e-2,-5e-2),(-5e-2,-5e-2)), ((1e-1,0),(0,1e-1),(-1e-1,0),(0,-1e-1)), ((1e-1,1e-1),(-1e-1,1e-1),(1e-1,-1e-1),(-1e-1,-1e-1)),((5e-1,0),(0,5e-1),(-5e-1,0),(0,-5e-1)), ((5e-1,2e-1),(-5e-1,2e-1),(5e-1,-2e-1),(-5e-1,-2e-1))]
for m, pm in zip(mismatch, posible_matches):
  lat, lng = m[-2:]
  zcode = m[-2:]
  for points in area:
    if len(pm.keys()) == 0: 
      print('TRY MISSING', d_lat , d_lng, 'at index', index)
      p_matches = dict(pm)
      for d_lat, d_lng in points:
        try:
          case = bing_data(lat + d_lat , lng + d_lng)
          p_matches.setdefault(case, 0)
          p_matches[case] += 1
          if case[-1] == zcode:
            break
        except Exception as e:
          print('Fail', d_lat , d_lng, 'at index', index)
      posible_matches[index] = p_matches
      pm = p_matches
  index += 1

TRY MISSING -0.01 -0.01 at index 31
Fail 0.005 0 at index 31
Fail 0 0.005 at index 31
Fail -0.005 0 at index 31
Fail 0 -0.005 at index 31
TRY MISSING 0 -0.005 at index 31
Fail 0.005 0.002 at index 31
Fail -0.005 0.002 at index 31
Fail 0.005 -0.002 at index 31
Fail -0.005 -0.002 at index 31
TRY MISSING -0.005 -0.002 at index 31
Fail 0 0.05 at index 31
Fail -0.05 0 at index 31
Fail 0 -0.05 at index 31
TRY MISSING 0 -0.05 at index 35
Fail 0.005 0 at index 35
Fail 0 0.005 at index 35
Fail -0.005 0 at index 35
Fail 0 -0.005 at index 35
TRY MISSING 0 -0.005 at index 35
Fail 0.005 0.002 at index 35
Fail -0.005 0.002 at index 35
Fail 0.005 -0.002 at index 35
Fail -0.005 -0.002 at index 35
TRY MISSING -0.005 -0.002 at index 35
Fail 0.05 0 at index 35
Fail -0.05 0 at index 35
Fail 0 -0.05 at index 35
TRY MISSING 0 -0.05 at index 50
Fail 0.005 0 at index 50
Fail 0 0.005 at index 50
Fail -0.005 0 at index 50
Fail 0 -0.005 at index 50
TRY MISSING 0 -0.005 at index 50
Fail 0.005 0.002 at index 50
Fa

In [ ]:
print('missing cases:',len([k for k in posible_matches if len(k.keys()) == 0]))
[m for m, pm in zip(mismatch, posible_matches) if len(pm.keys()) == 0]

missing cases: 0


[]

In [ ]:
#NOW WE ADD TO OUR USPS STRUCTURE THE NEW ZIPCODES THAT WE FOUND WITH BING
zipcodes_posible_matches = {zc[-1] for p in posible_matches for zc in p.keys()}
print('missing zipcodes', zipcodes_posible_matches - zipcode_data.keys())

# FIND MISSING ZIPCODES
for zipcode in sorted(zipcodes_posible_matches - zipcode_data.keys()):
  if zipcode not in zipcode_to_city.keys():
    zipcode_to_city.setdefault(zipcode, {})
    location = usps_zipcode_api(str(zipcode))
    print('missing', zipcode, location)
    for state in location.keys():
      zipcode_to_city[zipcode].setdefault(state, set())
      for city in location[state]:
        zipcode_to_city[zipcode][state].add(city)
        city_to_zipcode.setdefault(state, {})
        city_to_zipcode[state].setdefault(city, set())
        city_to_zipcode[state][city].add(zipcode)
  print(zipcode, zipcode_to_city[zipcode])
  

missing zipcodes {'98925', '27964', '96063', '04777', '06264', '95430', '23336', '95686', '71744', '63565', '04622', '36528', '88052', '04462', '10162', '04733', '70036', '93642', '47917', '04654', '04628', '46797'}
04462 {'ME': {'CEDAR LAKE', 'LONG A TWP', 'INDIAN PURCH', 'INDIAN PURCHASE TWP', 'MILLINOCKET', 'CEDAR LAKE TWP'}}
04622 {'ME': {'CHERRYFIELD', 'DEBLOIS', 'BEDDINGTON'}}
04628 {'ME': {'MARION TWP', 'DENNYSVILLE', 'EDMUNDS TWP'}}
04654 {'ME': {'ROQUE BLUFFS', 'NORTHFIELD', 'MARSHFIELD', 'DAY BLOCK TWP', 'MACHIAS', 'WHITNEYVILLE'}}
04733 {'ME': {'BENEDICTA'}}
04777 {'ME': {'STACYVILLE', 'HRSYTOWN TWP', 'HERSEYTOWN TWP', 'SHERMAN STA', 'SHERMAN STATION'}}
06264 {'CT': {'SCOTLAND'}}
10162 {'NY': {'NY CITY', 'NEW YORK CITY', 'NYC', 'MANHATTAN', 'NEW YORK', 'NY'}}
23336 {'VA': {'CHINCOTEAGUE ISLAND', 'CHINCOTEAGUE'}}
27964 {'NC': {'POINT HARBOR'}}
36528 {'AL': {'DAUPHIN ISLAND', 'DAUPHIN ISL'}}
46797 {'IN': {'EDGERTON', 'WOODBURN', 'MAUMEE'}}
47917 {'IN': {'AMBIA'}}
63565 {'MO': 

In [ ]:
# SAVE THE FILES FOR FUTURE COMPARISON
with open('/content/drive/My Drive/IA-TP1/zipcode_to_city.txt', 'w') as f:
	f.write(str(zipcode_to_city))
with open('/content/drive/My Drive/city_to_zipcode.txt', 'w') as f:
	f.write(str(city_to_zipcode))

In [ ]:

# SAVE THE FILES FOR FUTURE COMPARISON
with open('/content/drive/My Drive/mismatch.txt', 'w') as f:
	f.write(str(mismatch))
with open('/content/drive/My Drive/posible_matches.txt', 'w') as f:
	f.write(str(posible_matches))
 

In [ ]:
mismatch = eval(open('/content/drive/My Drive/mismatch.txt').read().replace('nan,', 'np.nan,'))
posible_matches = eval(open('/content/drive/My Drive/posible_matches.txt').read())
city_to_zipcode = eval(open('/content/drive/My Drive/city_to_zipcode.txt').read())
zipcode_to_city = eval(open('/content/drive/My Drive/zipcode_to_city.txt').read())

In [ ]:
#BEST MATCH ALGORITHM
best_match = []
for m, pm in zip(mismatch, posible_matches):
  zipcode, city, county, state, lat, lng = m
  possibles_states = {}
  possibles_counties = {}
  possibles_zipcodes = {}
  for k,v in pm.items():
    possibles_states.setdefault(k[0].upper(), 0)
    possibles_states[k[0].upper()] += v
    possibles_counties.setdefault(k[1].upper(), 0)
    possibles_counties[k[1].upper()] += v
    possibles_zipcodes.setdefault(k[2].upper(), 0)
    possibles_zipcodes[k[2].upper()] += v
  possibles_cities = {}
  for zc,f in possibles_zipcodes.items():
    for pstate in zipcode_to_city[zc].keys():
      possibles_cities.setdefault(pstate, {})
      for pcity in zipcode_to_city[zc][pstate]:
        possibles_cities[pstate].setdefault(pcity, [set(), 0])
        possibles_cities[pstate][pcity][0].add(zc)
        possibles_cities[pstate][pcity][1] += f
  pzipcode = ''
  pcity = ''
  pcounty = ''
  pstate = ''
  # ZIPCODE INPUT IS CORRECT
  if zipcode in possibles_zipcodes.keys(): # CORRECT ZIP
    pzipcode = zipcode
    pstate = list(zipcode_to_city[pzipcode].keys())[0] # IS UNIQUE BECAUSE OF ZIPCODE
    zipcode_possible_cities = zipcode_to_city[pzipcode][pstate]
    zipcode_pm = {k for k in pm.keys() if k[2] == pzipcode and k[0] == pstate} # KEEP ONLY ZIPCODE & STATE VIABLE
    zipcode_possibles_counties = {k[1].upper(): possibles_counties[k[1].upper()] for k in zipcode_pm}
    
    if len(zipcode_possibles_counties.keys()) == 1 and list(zipcode_possibles_counties.keys())[0] is not np.nan: # IF THE SOLUTION IS UNIQUE
      pcounty = list(possibles_counties.keys())[0].upper()
    elif county in possibles_counties.keys():  # CORRECT COUNTY
      pcounty = county.upper()
    else:
      pcounty = [k.upper() for k,v in possibles_counties.items() if v == max(zipcode_possibles_counties.values())][0]
    if len(zipcode_possible_cities) == 1 and list(zipcode_possible_cities)[0] is not np.nan: # IF THE SOLUTION IS UNIQUE
      pcity = list(zipcode_possible_cities)[0]
    elif city in zipcode_possible_cities: # CORRECT CITY
      pcity = city
    else:
      # TRY FINDING IF CITY MATCHES USPS DATA
      bing_state, bing_county, bing_city, bing_zipcode = '', '', '', ''
      try:
        bing_state, bing_county, bing_city, bing_zipcode = bing_data(lat, lng, use_locality=True)
        print(bing_state, bing_county, bing_city, bing_zipcode)
      except:
        pass
      #if bing_state != '':
      #  print(bing_state.upper(), pstate, bing_county.upper(), pcounty, bing_zipcode.upper(), pzipcode, bing_city.upper(), possibles_cities[pstate].keys())
      #  print()
      if bing_state.upper() == pstate and bing_county.upper() == pcounty and bing_zipcode.upper() == pzipcode and bing_city.upper() in possibles_cities[pstate].keys():
        pcity = bing_city.upper()
        print('BING CITY IMPROVEMENT')
      else:
        freq_c = {c.upper() : possibles_cities[pstate][c][1] for c in zipcode_possible_cities if c in possibles_cities[pstate]}
        max_freq = max(freq_c.values())
        pcity = [k for k,v in freq_c.items() if v == max_freq][0]
  
  
  else: # INVALID ZIPCODE
    if len(possibles_counties.keys()) == 1 or county in possibles_counties.keys():
      if len(possibles_counties.keys()) == 1 and list(possibles_counties.keys())[0] is not np.nan: # IF THE SOLUTION IS UNIQUE
        pcounty = list(possibles_counties.keys())[0].upper()
      elif county in possibles_counties.keys():  # CORRECT COUNTY
        pcounty = county.upper()
      pm = {k:v for k,v in pm.items() if k[1].upper() == pcounty}
      possibles_states = {}
      possibles_counties = {}
      possibles_zipcodes = {}
      for k,v in pm.items():
        possibles_states.setdefault(k[0].upper(), 0)
        possibles_states[k[0].upper()] += v
        possibles_counties.setdefault(k[1].upper(), 0)
        possibles_counties[k[1].upper()] += v
        possibles_zipcodes.setdefault(k[2].upper(), 0)
        possibles_zipcodes[k[2].upper()] += v
      
    if len(possibles_states.keys()) == 1 or state in possibles_states.keys():
      if len(possibles_states.keys()) == 1 and list(possibles_states.keys())[0] is not np.nan: # IF THE SOLUTION IS UNIQUE
        pstate = list(possibles_states.keys())[0].upper()
      elif state in possibles_states.keys():  # CORRECT STATE
        pstate = state.upper()
      pm = {k:v for k,v in pm.items() if k[0].upper() == pstate}
      possibles_states = {}
      possibles_counties = {}
      possibles_zipcodes = {}
      for k,v in pm.items():
        possibles_states.setdefault(k[0].upper(), 0)
        possibles_states[k[0].upper()] += v
        possibles_counties.setdefault(k[1].upper(), 0)
        possibles_counties[k[1].upper()] += v
        possibles_zipcodes.setdefault(k[2].upper(), 0)
        possibles_zipcodes[k[2].upper()] += v 
    
    possibles_cities = {}
    possibles_cities_to_filter = {}
    for zc,f in possibles_zipcodes.items():
      for ps in zipcode_to_city[zc].keys():
        possibles_cities.setdefault(ps,{})
        for pc in zipcode_to_city[zc][ps]:
          possibles_cities[ps].setdefault(pc, [set(), 0])
          possibles_cities[ps][pc][0].add(zc)
          possibles_cities[ps][pc][1] += f
          possibles_cities_to_filter.setdefault(pc, {'state':set(), 'zipcode': set()})
          possibles_cities_to_filter[pc]['state'].add(ps.upper())
          possibles_cities_to_filter[pc]['zipcode'].add(zc.upper())
    
    if city in possibles_cities_to_filter.keys():
      possibles_states_filter = possibles_cities_to_filter[city]['state']
      possibles_zipcode_filter = possibles_cities_to_filter[city]['zipcode']
      pm_filter = {k:v for k,v in pm.items() if k[0].upper() in  possibles_states_filter and k[2].upper() in  possibles_zipcode_filter}
      if len(pm_filter) > 1: # AVOID NULL ANSWER
        pm = pm_filter

    pm = {k:v for k,v in pm.items()}
    possibles_states = {}
    possibles_counties = {}
    possibles_zipcodes = {}
    for k,v in pm.items():
      possibles_states.setdefault(k[0].upper(), 0)
      possibles_states[k[0].upper()] += v
      possibles_counties.setdefault(k[1].upper(), 0)
      possibles_counties[k[1].upper()] += v
      possibles_zipcodes.setdefault(k[2].upper(), 0)
      possibles_zipcodes[k[2].upper()] += v 
    
    possibles_cities = {}
    for zc,f in possibles_zipcodes.items():
      for ps in zipcode_to_city[zc].keys():
        possibles_cities.setdefault(ps,{})
        for pc in zipcode_to_city[zc][ps]:
          possibles_cities[ps].setdefault(pc, [set(), 0])
          possibles_cities[ps][pc][0].add(zc)
          possibles_cities[ps][pc][1] += f
    
    if len(possibles_zipcodes.keys()) == 1 and list(possibles_zipcodes.keys())[0] is not np.nan:
      pzipcode = list(possibles_zipcodes.keys())[0]
    else:
      max_freq_pzipcode = max(possibles_zipcodes.values())
      pzipcode = [k for k,v in possibles_zipcodes.items() if v == max_freq_pzipcode][0]

    pstate = list(zipcode_to_city[pzipcode].keys())[0] # IS UNIQUE BECAUSE OF ZIPCODE
    zipcode_possible_cities = zipcode_to_city[pzipcode][pstate]
    zipcode_pm = {k for k in pm.keys() if k[2] == pzipcode and k[0] == pstate} # KEEP ONLY ZIPCODE & STATE VIABLE
    zipcode_possibles_counties = {k[1].upper(): possibles_counties[k[1].upper()] for k in zipcode_pm}
    
    if pcounty == '':
      if len(zipcode_possibles_counties.keys()) == 1 and list(zipcode_possibles_counties.keys())[0] is not np.nan: # IF THE SOLUTION IS UNIQUE
        pcounty = list(possibles_counties.keys())[0].upper()
      elif county in possibles_counties.keys():  # CORRECT COUNTY
        pcounty = county.upper()
      else:
        pcounty = [k.upper() for k,v in possibles_counties.items() if v == max(zipcode_possibles_counties.values())][0]
    if pcity == '':    
      if len(zipcode_possible_cities) == 1 and list(zipcode_possible_cities)[0] is not np.nan:  # IF THE SOLUTION IS UNIQUE
        pcity = list(zipcode_possible_cities)[0]
      elif city in zipcode_possible_cities: # CORRECT CITY
        pcity = city
      else:
        # TRY FINDING IF CITY MATCHES USPS DATA
        bing_state, bing_county, bing_city, bing_zipcode = '', '', '', ''
        try:
          bing_state, bing_county, bing_city, bing_zipcode = bing_data(lat, lng, use_locality=True)
          print(bing_state, bing_county, bing_city, bing_zipcode)
        except:
          pass
        #if bing_state != '':
        #  print(bing_state.upper(), pstate, bing_county.upper(), pcounty, bing_zipcode.upper(), pzipcode, bing_city.upper(), possibles_cities[pstate].keys())
        #  print()
        if bing_state.upper() == pstate and bing_county.upper() == pcounty and bing_zipcode.upper() == pzipcode and bing_city.upper() in possibles_cities[pstate].keys():
          pcity = bing_city.upper()
          print("\033[0;33mBING CITY IMPROVEMENT\033[0m")
        else:
          freq_c = {c.upper() : possibles_cities[pstate][c][1] for c in zipcode_possible_cities if c in possibles_cities[pstate]}
          max_freq = max(freq_c.values())
          pcity = [k for k,v in freq_c.items() if v == max_freq][0]
  print('MISMATCH:', "\033[0;31m" + str(m) + "\033[0m", 'CORRECT:',"\033[0;32m" + str((pzipcode, pcity, pcounty, pstate)) + "\033[0m")
  best_match.append((pzipcode, pcity, pcounty, pstate))

MISMATCH: (nan, 'SOUTHEAST MARIN', 'MARIN', 'CA', 37.885376, -122.516586) CORRECT: ('94941', 'STRAWBERRY POINT', 'MARIN', 'CA')
MISMATCH: ('95616', 'DIXON', 'SOLANO', 'CA', 38.536564, -121.74000500000001) CORRECT: ('95616', 'DAVIS', 'SOLANO', 'CA')
MISMATCH: ('95616', 'DIXON', 'SOLANO', 'CA', 38.536564, -121.74000500000001) CORRECT: ('95616', 'DAVIS', 'SOLANO', 'CA')
MISMATCH: (nan, 'SOUTHEAST MARIN', 'MARIN', 'CA', 37.885376, -122.516586) CORRECT: ('94941', 'STRAWBERRY POINT', 'MARIN', 'CA')
MISMATCH: (nan, 'SOUTHEAST MARIN', 'MARIN', 'CA', 37.885376, -122.516586) CORRECT: ('94941', 'STRAWBERRY POINT', 'MARIN', 'CA')
MISMATCH: (nan, 'SOUTHEAST MARIN', 'MARIN', 'CA', 37.885376, -122.516586) CORRECT: ('94941', 'STRAWBERRY POINT', 'MARIN', 'CA')
MISMATCH: (nan, 'SAN MATEO', 'SAN MATEO', 'CA', 37.505482, -122.11930100000001) CORRECT: ('94025', 'WEST MENLO PARK', 'SAN MATEO', 'CA')
MISMATCH: (nan, 'SOUTHEAST MARIN', 'MARIN', 'CA', 37.885376, -122.516586) CORRECT: ('94941', 'STRAWBERRY POIN

In [ ]:
# SAVE THE FILES FOR FUTURE COMPARISON
with open('/content/drive/My Drive/mismatch.txt', 'w') as f:
	f.write(str(mismatch))
with open('/content/drive/My Drive/posible_matches.txt', 'w') as f:
	f.write(str(posible_matches))
with open('/content/drive/My Drive/best_match.txt', 'w') as f:
	f.write(str(best_match))
with open('/content/drive/My Drive/city_to_zipcode.txt', 'w') as f:
	f.write(str(city_to_zipcode))
with open('/content/drive/My Drive/zipcode_to_city.txt', 'w') as f:
	f.write(str(zipcode_to_city))
 

In [ ]:
mismatch = eval(open('/content/drive/My Drive//mismatch.txt').read().replace('nan,', 'np.nan,'))
posible_matches = eval(open('/content/drive/My Drive/posible_matches.txt').read())
city_to_zipcode = eval(open('/content/drive/My Drive/city_to_zipcode.txt').read())
zipcode_to_city = eval(open('/content/drive/My Drive/zipcode_to_city.txt').read())
best_match = eval(open('/content/drive/My Drive/best_match.txt').read())

In [ ]:
#WE CREATE THE MODIFIED COLUMNS TO UPDATE OUR DATASET
zipcode_correct_values = []
city_correct_values = []
county_correct_values = []
state_correct_values = []
lat_correct_values = []
lng_correct_values = []
best_match_index = 0

for zipcode, city, county, state, lat, lng in zip (us_accident_data.Zipcode, us_accident_data.City, us_accident_data.County, us_accident_data.State, us_accident_data.Start_Lat, us_accident_data.Start_Lng):
  if not(zipcode in zipcode_to_city.keys() and state in zipcode_to_city[zipcode].keys() and city in zipcode_to_city[zipcode][state] and county is not np.nan):
    if (mismatch[best_match_index] == (zipcode, city, county, state, lat, lng)):
      zipcode, city, county, state = best_match[best_match_index]
      best_match_index +=1
    else:
      print("\033[0;31m ERROR \033[0m")
  zipcode_correct_values.append(zipcode)
  city_correct_values.append(city)
  county_correct_values.append(county)
  state_correct_values.append(state)
  lat_correct_values.append(lat)
  lng_correct_values.append(lng)



In [ ]:
#CHECK SOLUTION
assert all(lat_correct_values == us_accident_data.Start_Lat), 'Incorrect solution'
assert all(lng_correct_values == us_accident_data.Start_Lng), 'Incorrect solution'

In [ ]:
# ASSIGN BEST MATCH
us_accident_data.Zipcode = zipcode_correct_values
us_accident_data.City = city_correct_values
us_accident_data.County = county_correct_values
us_accident_data.State = state_correct_values

In [ ]:
#CHECK ASSIGNMENT
assert all(zipcode_correct_values == us_accident_data.Zipcode), 'Incorrect assignment'
assert all(city_correct_values == us_accident_data.City), 'Incorrect assignment'
assert all(county_correct_values == us_accident_data.County), 'Incorrect assignment'
assert all(state_correct_values == us_accident_data.State), 'Incorrect assignment'

In [ ]:
# SAVE PROGRESS
us_accident_data.to_csv('/content/drive/My Drive/us_accident_data_after_location_update.csv')

In [ ]:
# LOAD PROGRESS
us_accident_data = pd.read_csv('/content/drive/My Drive/us_accident_data_after_location_update.csv')

In [ ]:
#POSSIBLE VALUES TIMEZONES
print('Possible Timezone with format', set(us_accident_data.Timezone.unique()) - {np.nan})

Possible Timezone with format {'US/Eastern', 'US/Mountain', 'US/Central', 'US/Pacific'}


### Timezone analysis

Since we change some data and update the zipcodes, NANs, we proceed to make verification with the BING API to check if it is consistent with what we have, making requests from the Postal Code

In [ ]:
import requests, json

def bing_timezone_api(lat, lng):
  headers = {
    'authority': 'dev.virtualearth.net',
    'pragma': 'no-cache',
    'cache-control': 'no-cache',
    'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36',
    'dnt': '1',
    'accept': '*/*',
    'sec-fetch-site': 'cross-site',
    'sec-fetch-mode': 'no-cors',
    'sec-fetch-dest': 'script',
    'referer': 'https://dev.virtualearth.net/',
    'accept-language': 'en-US,en;q=0.9',
  }

  params = (
      ('o', 'json'),
      ('jsonp', 'Microsoft.Maps.NetworkCallbacks.f77b6a'),
      ('key', 'AoltDrgPXkLphU6xf7mwndXly_XEVrnForv0FrqgPHHXwIf-oaB3bfErNvlSYVXc'),
      ('includeEntityTypes', 'Neighborhood,PopulatedPlace,AdminDivision1,AdminDivision2,CountryRegion'),
      ('inclnb', '1'),
      ('incl', 'entityId'),
      ('c', 'en-US')
  )
  response = requests.get(f'https://dev.virtualearth.net/REST/v1/TimeZone/{lat},{lng}', headers=headers, params=params)
  data =  json.loads(response.text[response.text.find('{'):response.text.rfind('}') + 1])
  timezones = [{k:v} for resourceSets in data['resourceSets'] for resources in resourceSets['resources'] for k, v in resources['timeZone'].items() if k in {'genericName', 'abbreviation', 'windowsTimeZoneId'}]
  
  timezone = set()
  timezone_format = {'Eastern Standard Time': 'EST', 'Central Standard Time' : 'CST' , 'Mountain Standard Time' : 'MST', 'Pacific Standard Time' : 'PST'}
  for t in timezones:
    if 'abbreviation' in t.keys():
      timezone.add(t['abbreviation'])
    elif 'genericName' in t.keys() and t['genericName'].find('Standard Time') > 0:
      for k,v in timezone_format.items():
        if t['genericName'].find(k) >= 0:
          timezone.add(v)
          break
    elif 'windowsTimeZoneId' in t.keys() and t['windowsTimeZoneId'].find('Standard Time') > 0:
      for k,v in timezone_format.items():
        if t['windowsTimeZoneId'].find(k) >= 0:
          timezone.add(v)
          break
    else:
      print(f'ERROR TIMEZONE {timezone}')
      raise f'ERROR TIMEZONE {timezone}'

  return sorted(timezone)

In [ ]:
import requests, json
def bing_timezone_api_by_zipcode(zipcode):
  headers = {
  'authority': 'dev.virtualearth.net',
  'pragma': 'no-cache',
  'cache-control': 'no-cache',
  'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36',
  'dnt': '1',
  'accept': '*/*',
  'sec-fetch-site': 'cross-site',
  'sec-fetch-mode': 'no-cors',
  'sec-fetch-dest': 'script',
  'referer': 'https://dev.virtualearth.net/',
  'accept-language': 'en-US,en;q=0.9',
  }

  params = (
      ('o', 'json'),
      ('jsonp', 'Microsoft.Maps.NetworkCallbacks.normal'),
      ('jsonso', '394594'),
      ('key', 'AoltDrgPXkLphU6xf7mwndXly_XEVrnForv0FrqgPHHXwIf-oaB3bfErNvlSYVXc'),
      ('inclnb', '1'),
      ('incl', 'entityId'),
      ('c', 'en-US'),
      ('q', zipcode)
      
  )

  response = requests.get(f'https://dev.virtualearth.net/REST/v1/TimeZone', headers=headers, params=params)
  data =  json.loads(response.text[response.text.find('{'):response.text.rfind('}') + 1])

  timezones = [time_zone for resourceSets in data['resourceSets'] for resources in resourceSets['resources'] for time_zone_at_location in resources['timeZoneAtLocation'] for time_zone in time_zone_at_location['timeZone']]

  timezone = set()
  timezone_format = {'Eastern Standard Time': 'EST', 'Central Standard Time' : 'CST' , 'Mountain Standard Time' : 'MST', 'Pacific Standard Time' : 'PST'}
  for t in timezones:
    if 'abbreviation' in t.keys():
      timezone.add(t['abbreviation'])
    elif 'genericName' in t.keys() and t['genericName'].find('Standard Time') > 0:
      for k,v in timezone_format.items():
        if t['genericName'].find(k) >= 0:
          timezone.add(v)
          break
    elif 'windowsTimeZoneId' in t.keys() and t['windowsTimeZoneId'].find('Standard Time') > 0:
      for k,v in timezone_format.items():
        if t['windowsTimeZoneId'].find(k) >= 0:
          timezone.add(v)
          break
    else:
      print(f'ERROR TIMEZONE {timezone}')
      raise f'ERROR TIMEZONE {timezone}'

  return sorted(timezone)

In [ ]:
all_zipcodes = sorted({zc for zc in us_accident_data.Zipcode})
zipcode_to_timezone = {}
format_timezone = {'EST' : 'US/Eastern', 'CST': 'US/Central', 'MST': 'US/Mountain', 'PST': 'US/Pacific'}
fail_zipcodes_timezone_request = set()
multiple_zipcode_timezone_request = set()

In [ ]:
all_zipcodes = sorted({zc for zc in us_accident_data.Zipcode})
for zipcode in all_zipcodes:
  if zipcode not in zipcode_to_timezone.keys():
    try:
      zipcode_timezone = bing_timezone_api_by_zipcode(zipcode)
      if len(zipcode_timezone) == 1:
        zipcode_to_timezone[zipcode] = zipcode_timezone[0]
        print(f"\033[0;32m{zipcode}: {zipcode_to_timezone[zipcode]}\033[0m")
      elif len(zipcode_timezone) > 1:
        print(f"\033[0;33m{zipcode}: {zipcode_timezone}\033[0m")
        multiple_zipcode_timezone_request.add(zipcode)
      else:
        print(f"\033[0;31m{zipcode}: FAIL\033[0m")
        fail_zipcodes_timezone_request.add(zipcode)
    except:
      print(f"\033[0;31m{zipcode}: FAIL\033[0m")
      fail_zipcodes_timezone_request.add(zipcode)
  else:
    print(f"\033[0;34m{zipcode}: {zipcode_to_timezone[zipcode]}\033[0m")
    pass


Streaming output truncated to the last 5000 lines.
75432: CST
75433: CST
75440: CST
75442: CST
75452: CST
75453: CST
75454: CST
75455: CST
75457: CST
75459: CST
75460: CST
75462: CST
75469: CST
75470: CST
75473: CST
75474: CST
75478: CST
75482: CST
75491: CST
75493: CST
75494: CST
75495: CST
75501: CST
75503: CST
75551: CST
75558: CST
75559: CST
75561: CST
75563: CST
75567: CST
75568: CST
75570: CST
75571: CST
75572: CST
75574: CST
75601: CST
75602: CST
75603: CST
75604: CST
75605: CST
75630: CST
75631: CST
75638: CST
75639: CST
75644: CST
75645: CST
75650: CST
75652: CST
75657: CST
75662: CST
75669: CST
75670: CST
75672: CST
75683: CST
75684: CST
75692: CST
75701: CST
75702: CST
75703: CST
75704: CST
75705: CST
75706: CST
75707: CST
75708: CST
75709: CST
75751: CST
75752: CST
75763: CST
75766: CST
75771: CST
75773: CST
75783: CST
75785: CST
75790: CST
75792: CST
75801: CST
75803: CST
75831: CST
75833: CST
75835: CST
75840: CST
75844: CST
75845: CST
75847: CST
75850: CST
75851: CST
758

In [ ]:
# FIND MISSING BY LAT / LONG
# IN SOME REQUESTS BING FAILED TO FECTH THE DATA, THAT'S WHY WE MADE QUERIES BY LAT AND LNG

for zipcode, lat, lng in zip(us_accident_data.Zipcode, us_accident_data.Start_Lat, us_accident_data.Start_Lng):
  if zipcode not in zipcode_to_timezone.keys():
    try:
      zipcode_timezone = bing_timezone_api(lat, lng)
      if len(zipcode_timezone) == 1:
        zipcode_to_timezone[zipcode] = zipcode_timezone[0]
        print(f"\033[0;32m{zipcode}: {zipcode_to_timezone[zipcode]}\033[0m")
      else:
        print(f"\033[0;33m{zipcode}-{lat},{lng}: {zipcode_timezone}\033[0m")
    except:
      print(f"\033[0;31m{zipcode}-{lat},{lng}: FAIL\033[0m")

44136: EST
45171: EST
90640: PST
90069: PST
68059: CST
60137: CST
63304: CST
63124: CST
63128: CST
63127: CST
63123: CST
53144: CST
63130: CST
63376: CST
60153: CST
60070: CST
63135: CST
49444: EST
46507: EST
10901: EST
11706: EST
11209: EST
12603: EST
12522: EST
12545: EST
12514: EST
12508: EST
12550: EST
12033: EST
10801: EST
12306: EST
12531: EST
11362: EST
11003: EST
06471: EST
12061: EST
11001: EST
11763: EST
12017: EST
11590: EST
11375: EST
10970: EST
12140: EST
11722: EST
12052: EST
11561: EST
12580: EST
06512: EST
10001: EST
10502: EST
12018: EST
11776: EST
11105: EST
12198: EST
10804: EST
12157: EST
11520: EST
11212: EST
11501: EST
11208: EST
12054: EST
02738: EST
12515: EST
44830: EST
43410: EST
15101: EST
21122: EST
19335: EST
17025: EST
20111: EST
15001: EST
78109: CST
46540: EST
10993: EST
11581: EST
98361: PST
12019: EST
41015: EST
55432: CST
27617: EST
27615: EST
80524: MST
80111: MST
85138: MST
85251: MST
12020: EST
18640: EST
80104: MST
41017: EST
88339: MST
14001: EST

In [ ]:
print('missing zipcode timezone:', set(us_accident_data.Zipcode) - zipcode_to_timezone.keys())

missing zipcode timezone: set()


In [ ]:
# SAVE PROGRESS
with open('/content/drive/My Drive/IA-TP1/zipcode_to_timezone.txt', 'w') as f:
	f.write(str(zipcode_to_timezone))
with open('/content/drive/My Drive/IA-TP1/fail_zipcodes_timezone_request.txt', 'w') as f:
	f.write(str(fail_zipcodes_timezone_request))
with open('/content/drive/My Drive/IA-TP1/multiple_zipcode_timezone_request.txt', 'w') as f:
	f.write(str(multiple_zipcode_timezone_request))

In [ ]:
# LOAD PROGRESS
format_timezone = {'EST' : 'US/Eastern', 'CST': 'US/Central', 'MST': 'US/Mountain', 'PST': 'US/Pacific'}
zipcode_to_timezone = eval(open('/content/drive/My Drive/IA-TP1/zipcode_to_timezone.txt').read())
fail_zipcodes_timezone_request = eval(open('/content/drive/My Drive/IA-TP1/fail_zipcodes_timezone_request.txt').read())
multiple_zipcode_timezone_request = eval(open('/content/drive/My Drive/IA-TP1/multiple_zipcode_timezone_request.txt').read())

### Converting _start_ _time_ to  UTC + 0 

In [ ]:
from datetime import datetime, timedelta
timezone_to_utc = {'EST': timedelta(hours=-5), 'CST': timedelta(hours=-6), 'MST': timedelta(hours=-7), 'PST': timedelta(hours=-8)}
zipcode_to_utc = {zipcode: timezone_to_utc[timezone] for zipcode, timezone in zipcode_to_timezone.items()}
us_accident_data.Start_Time = pd.DatetimeIndex(us_accident_data.Start_Time)
us_accident_data.Start_Time = [time + zipcode_to_utc[zipcode] for time, zipcode in zip(us_accident_data.Start_Time, us_accident_data.Zipcode)]

In [ ]:
us_accident_data.Timezone = [zipcode_to_timezone[zipcode] for zipcode in us_accident_data.Zipcode]

In [ ]:
# SAVE PROGRESS
us_accident_data.to_csv('/content/drive/My Drive/us_accident_data_timezone_updated.csv')

In [ ]:
# LOAD PROGRESS
us_accident_data = pd.read_csv('/content/drive/My Drive/us_accident_data_timezone_updated.csv')

#### Filling _wind_ _speed_ and  _wind_ _direction_ 

If not possible we will use the data from state, county, and city

In [ ]:
wind_direction = us_accident_data.Wind_Direction
wind_speed = us_accident_data['Wind_Speed(mph)']
us_accident_data['Wind_Speed(mph)'] = [wind_speed if not np.isnan(wind_speed) else [0 if wind_direction == 'Calm' else wind_speed][0] for wind_speed, wind_direction in zip(us_accident_data['Wind_Speed(mph)'], us_accident_data.Wind_Direction)]

In [ ]:
# WEATHER CONDITION ANALYSIS TAKING INTO ACCOUNT NAN WIND SPEED
x = us_accident_data['Wind_Speed(mph)']
from collections import Counter
weather_conditions = set(us_accident_data.Weather_Condition)
weather_conditions_nan_speed_wind = Counter(us_accident_data.Weather_Condition[np.isnan(x)])
weather_conditions_notnan_speed_wind = Counter(us_accident_data.Weather_Condition[np.isnan(x)==False])
loss_percentage = sorted([(key, 100*weather_conditions_nan_speed_wind.get(key, 0)/(weather_conditions_notnan_speed_wind.get(key, 0) + weather_conditions_nan_speed_wind.get(key, 0))) for key in weather_conditions], key=lambda x: -x[1])

print('WEATHER CONDITION CASES WITH NO SPEED')
for key, percentage in loss_percentage:
  print(f"\033[0;34m{key}\033[0m", 'KEEP:', weather_conditions_notnan_speed_wind.get(key, 0), 'LOSE:',  weather_conditions_nan_speed_wind.get(key, 0), f"\033[0;31m{percentage:.2f}%\033[0m")


WEATHER CONDITION CASES WITH NO SPEED
Volcanic Ash KEEP: 0 LOSE: 22 100.00%
nan KEEP: 12128 LOSE: 53804 81.61%
Sand KEEP: 11 LOSE: 8 42.11%
N/A Precipitation KEEP: 187 LOSE: 14 6.97%
Light Snow Showers KEEP: 23 LOSE: 1 4.17%
Heavy Snow KEEP: 1201 LOSE: 48 3.84%
Mist KEEP: 2174 LOSE: 30 1.36%
Light Freezing Fog KEEP: 988 LOSE: 13 1.30%
Thunder KEEP: 1640 LOSE: 21 1.26%
Haze KEEP: 33901 LOSE: 414 1.21%
Heavy Drizzle KEEP: 255 LOSE: 3 1.16%
Thunder in the Vicinity KEEP: 2154 LOSE: 23 1.06%
Snow KEEP: 4749 LOSE: 47 0.98%
Thunderstorm KEEP: 4395 LOSE: 43 0.97%
Fair KEEP: 332370 LOSE: 2919 0.87%
Clear KEEP: 801237 LOSE: 6934 0.86%
T-Storm KEEP: 2144 LOSE: 17 0.79%
Blowing Snow KEEP: 266 LOSE: 2 0.75%
Heavy Thunderstorms and Rain KEEP: 2466 LOSE: 17 0.68%
Smoke KEEP: 3579 LOSE: 23 0.64%
Drizzle KEEP: 2031 LOSE: 13 0.64%
Rain KEEP: 32629 LOSE: 197 0.60%
Cloudy KEEP: 114815 LOSE: 681 0.59%
Heavy Rain KEEP: 11995 LOSE: 69 0.57%
Overcast KEEP: 380421 LOSE: 2059 0.54%
Light Snow KEEP: 41900 LOSE: 

In [ ]:
# FILL BY WEATHER CONDITION & SHORTER
threshold = 5
wind_speed_update = []
sample_filler = {}
for wind_speed, weather, zipcode, city, county, state in zip(us_accident_data['Wind_Speed(mph)'], us_accident_data.Weather_Condition, us_accident_data.Zipcode, us_accident_data.City, us_accident_data.County, us_accident_data.State):
  if not np.isnan(wind_speed) and type(weather) == str:
    # CONSIDERING WEATHER CONDITION
    sample_filler.setdefault(weather+state, [0,0])
    sample_filler.setdefault(weather+state+county, [0,0])
    sample_filler.setdefault(weather+state+county+city, [0,0])
    sample_filler.setdefault(weather+zipcode, [0,0])
    # ADD WINDSPEED
    # CONSIDERING WEATHER CONDITION
    sample_filler[weather+state][0] += wind_speed
    sample_filler[weather+state+county][0] += wind_speed
    sample_filler[weather+state+county+city][0] += wind_speed
    sample_filler[weather+zipcode][0] += wind_speed
    # COUNT CASES
    # CONSIDERING WEATHER CONDITION
    sample_filler[weather+state][1] += 1
    sample_filler[weather+state+county][1] += 1
    sample_filler[weather+state+county+city][1] += 1
    sample_filler[weather+zipcode][1] += 1
  if not np.isnan(wind_speed):  
    # CONSIDERING ONLY LOCATION
    sample_filler.setdefault(state, [0,0])
    sample_filler.setdefault(state+county, [0,0])
    sample_filler.setdefault(state+county+city, [0,0])
    sample_filler.setdefault(zipcode, [0,0])
    # ADD WINDSPEED
    # CONSIDERING ONLY LOCATION
    sample_filler[state][0] += wind_speed
    sample_filler[state+county][0] += wind_speed
    sample_filler[state+county+city][0] += wind_speed
    sample_filler[zipcode][0] += wind_speed
    # COUNT CASES
    # CONSIDERING ONLY LOCATION
    sample_filler[state][1] += 1
    sample_filler[state+county][1] += 1
    sample_filler[state+county+city][1] += 1
    sample_filler[zipcode][1] += 1

# FIND MEAN WHEN THE CASES ARE GREATER THAN THE THRESHOLD
mean_filler = {key:value[0]/value[-1] for key, value in sample_filler.items() if value[-1] >= threshold}


print()
print(f'STARTING CONDITION WITH THRESHOLD OF {threshold}')
print('nan cases', len(us_accident_data['Wind_Speed(mph)'][us_accident_data['Wind_Speed(mph)'].isna()]), 'nan percentage', 100 * len(us_accident_data['Wind_Speed(mph)'][us_accident_data['Wind_Speed(mph)'].isna()]) / len(us_accident_data['Wind_Speed(mph)']))
print()
for wind_speed, weather, zipcode, city, county, state in zip(us_accident_data['Wind_Speed(mph)'], us_accident_data.Weather_Condition, us_accident_data.Zipcode, us_accident_data.City, us_accident_data.County, us_accident_data.State):
  if np.isnan(wind_speed):
    if type(weather) == str: # TRY FILLING WITH WEATHER CONDITION
      wind_speed = mean_filler.get(weather+zipcode, np.nan) if np.isnan(wind_speed) else wind_speed
      wind_speed = mean_filler.get(weather+state+county+city, np.nan) if np.isnan(wind_speed) else wind_speed
      wind_speed = mean_filler.get(weather+state+county, np.nan) if np.isnan(wind_speed) else wind_speed
      wind_speed = mean_filler.get(weather+state, np.nan) if np.isnan(wind_speed) else wind_speed
    wind_speed = mean_filler.get(zipcode, np.nan) if np.isnan(wind_speed) else wind_speed
    wind_speed = mean_filler.get(state+county+city, np.nan) if np.isnan(wind_speed) else wind_speed
    wind_speed = mean_filler.get(state+county, np.nan) if np.isnan(wind_speed) else wind_speed
    wind_speed = mean_filler.get(state, np.nan) if np.isnan(wind_speed) else wind_speed
  wind_speed_update.append(wind_speed)
wind_speed_update_nan = len([xi for xi in wind_speed_update if np.isnan(xi)])
print('nan cases', wind_speed_update_nan, 'nan percentage', 100 * wind_speed_update_nan / len(wind_speed_update))
print()


STARTING CONDITION WITH THRESHOLD OF 5
nan cases 72572 nan percentage 2.439940356415804

nan cases 0 nan percentage 0.0



In [ ]:
assert all(np.isnan(us_accident_data['Wind_Speed(mph)'][us_accident_data['Wind_Speed(mph)'] != wind_speed_update])), 'INVALID CORRECTION'
assert not any(np.isnan(us_accident_data['Wind_Speed(mph)'][us_accident_data['Wind_Speed(mph)'] == wind_speed_update])), 'INVALID CORRECTION'

In [ ]:
us_accident_data['Wind_Speed(mph)'] = wind_speed_update

In [ ]:
# SAVE PROGRESS
us_accident_data.to_csv('/content/drive/My Drive/us_accident_data_wind_correction.csv')   

In [ ]:
# LOAD PROGRESS
us_accident_data = pd.read_csv('/content/drive/My Drive/us_accident_data_wind_correction.csv')

In [ ]:
# FILL BY WEATHER CONDITION & SHORTER
threshold = 5
temperature_update = []
sample_filler = {}
for temperature, weather, zipcode, city, county, state in zip(us_accident_data['Temperature(F)'], us_accident_data.Weather_Condition, us_accident_data.Zipcode, us_accident_data.City, us_accident_data.County, us_accident_data.State):
  if not np.isnan(temperature) and type(weather) == str:
    # CONSIDERING WEATHER CONDITION
    sample_filler.setdefault(weather+state, [0,0])
    sample_filler.setdefault(weather+state+county, [0,0])
    sample_filler.setdefault(weather+state+county+city, [0,0])
    sample_filler.setdefault(weather+zipcode, [0,0])
    # ADD WINDSPEED
    # CONSIDERING WEATHER CONDITION
    sample_filler[weather+state][0] += temperature
    sample_filler[weather+state+county][0] += temperature
    sample_filler[weather+state+county+city][0] += temperature
    sample_filler[weather+zipcode][0] += temperature
    # COUNT CASES
    # CONSIDERING WEATHER CONDITION
    sample_filler[weather+state][1] += 1
    sample_filler[weather+state+county][1] += 1
    sample_filler[weather+state+county+city][1] += 1
    sample_filler[weather+zipcode][1] += 1
  if not np.isnan(temperature):  
    # CONSIDERING ONLY LOCATION
    sample_filler.setdefault(state, [0,0])
    sample_filler.setdefault(state+county, [0,0])
    sample_filler.setdefault(state+county+city, [0,0])
    sample_filler.setdefault(zipcode, [0,0])
    # ADD WINDSPEED
    # CONSIDERING ONLY LOCATION
    sample_filler[state][0] += temperature
    sample_filler[state+county][0] += temperature
    sample_filler[state+county+city][0] += temperature
    sample_filler[zipcode][0] += temperature
    # COUNT CASES
    # CONSIDERING ONLY LOCATION
    sample_filler[state][1] += 1
    sample_filler[state+county][1] += 1
    sample_filler[state+county+city][1] += 1
    sample_filler[zipcode][1] += 1

# FIND MEAN WHEN THE CASES ARE GREATER THAN THE THRESHOLD
mean_filler = {key:value[0]/value[-1] for key, value in sample_filler.items() if value[-1] >= threshold}


print()
print(f'STARTING CONDITION WITH THRESHOLD OF {threshold}')
print('nan cases', len(us_accident_data['Temperature(F)'][us_accident_data['Temperature(F)'].isna()]), 'nan percentage', 100 * len(us_accident_data['Temperature(F)'][us_accident_data['Temperature(F)'].isna()]) / len(us_accident_data['Temperature(F)']))
print()
for temperature, weather, zipcode, city, county, state in zip(us_accident_data['Temperature(F)'], us_accident_data.Weather_Condition, us_accident_data.Zipcode, us_accident_data.City, us_accident_data.County, us_accident_data.State):
  if np.isnan(temperature):
    if type(weather) == str: # TRY FILLING WITH WEATHER CONDITION
      temperature = mean_filler.get(weather+zipcode, np.nan) if np.isnan(temperature) else temperature
      temperature = mean_filler.get(weather+state+county+city, np.nan) if np.isnan(temperature) else temperature
      temperature = mean_filler.get(weather+state+county, np.nan) if np.isnan(temperature) else temperature
      temperature = mean_filler.get(weather+state, np.nan) if np.isnan(temperature) else temperature
    temperature = mean_filler.get(zipcode, np.nan) if np.isnan(temperature) else temperature
    temperature = mean_filler.get(state+county+city, np.nan) if np.isnan(temperature) else temperature
    temperature = mean_filler.get(state+county, np.nan) if np.isnan(temperature) else temperature
    temperature = mean_filler.get(state, np.nan) if np.isnan(temperature) else temperature
  temperature_update.append(temperature)
temperature_update_nan = len([xi for xi in temperature_update if np.isnan(xi)])
print('nan cases', temperature_update_nan, 'nan percentage', 100 * temperature_update_nan / len(temperature_update))
print()


STARTING CONDITION WITH THRESHOLD OF 5
nan cases 56063 nan percentage 1.8848919170167449

nan cases 0 nan percentage 0.0



In [ ]:
us_accident_data['Temperature(F)'] = temperature_update

In [ ]:
# FILL BY WEATHER CONDITION & SHORTER
threshold = 5
humidity_update = []
sample_filler = {}
for humidity, weather, zipcode, city, county, state in zip(us_accident_data['Humidity(%)'], us_accident_data.Weather_Condition, us_accident_data.Zipcode, us_accident_data.City, us_accident_data.County, us_accident_data.State):
  if not np.isnan(humidity) and type(weather) == str:
    # CONSIDERING WEATHER CONDITION
    sample_filler.setdefault(weather+state, [0,0])
    sample_filler.setdefault(weather+state+county, [0,0])
    sample_filler.setdefault(weather+state+county+city, [0,0])
    sample_filler.setdefault(weather+zipcode, [0,0])
    # ADD WINDSPEED
    # CONSIDERING WEATHER CONDITION
    sample_filler[weather+state][0] += humidity
    sample_filler[weather+state+county][0] += humidity
    sample_filler[weather+state+county+city][0] += humidity
    sample_filler[weather+zipcode][0] += humidity
    # COUNT CASES
    # CONSIDERING WEATHER CONDITION
    sample_filler[weather+state][1] += 1
    sample_filler[weather+state+county][1] += 1
    sample_filler[weather+state+county+city][1] += 1
    sample_filler[weather+zipcode][1] += 1
  if not np.isnan(humidity):  
    # CONSIDERING ONLY LOCATION
    sample_filler.setdefault(state, [0,0])
    sample_filler.setdefault(state+county, [0,0])
    sample_filler.setdefault(state+county+city, [0,0])
    sample_filler.setdefault(zipcode, [0,0])
    # ADD WINDSPEED
    # CONSIDERING ONLY LOCATION
    sample_filler[state][0] += humidity
    sample_filler[state+county][0] += humidity
    sample_filler[state+county+city][0] += humidity
    sample_filler[zipcode][0] += humidity
    # COUNT CASES
    # CONSIDERING ONLY LOCATION
    sample_filler[state][1] += 1
    sample_filler[state+county][1] += 1
    sample_filler[state+county+city][1] += 1
    sample_filler[zipcode][1] += 1

# FIND MEAN WHEN THE CASES ARE GREATER THAN THE THRESHOLD
mean_filler = {key:value[0]/value[-1] for key, value in sample_filler.items() if value[-1] >= threshold}


print()
print(f'STARTING CONDITION WITH THRESHOLD OF {threshold}')
print('nan cases', len(us_accident_data['Humidity(%)'][us_accident_data['Humidity(%)'].isna()]), 'nan percentage', 100 * len(us_accident_data['Humidity(%)'][us_accident_data['Humidity(%)'].isna()]) / len(us_accident_data['Humidity(%)']))
print()
for humidity, weather, zipcode, city, county, state in zip(us_accident_data['Humidity(%)'], us_accident_data.Weather_Condition, us_accident_data.Zipcode, us_accident_data.City, us_accident_data.County, us_accident_data.State):
  if np.isnan(humidity):
    if type(weather) == str: # TRY FILLING WITH WEATHER CONDITION
      humidity = mean_filler.get(weather+zipcode, np.nan) if np.isnan(humidity) else humidity
      humidity = mean_filler.get(weather+state+county+city, np.nan) if np.isnan(humidity) else humidity
      humidity = mean_filler.get(weather+state+county, np.nan) if np.isnan(humidity) else humidity
      humidity = mean_filler.get(weather+state, np.nan) if np.isnan(humidity) else humidity
    humidity = mean_filler.get(zipcode, np.nan) if np.isnan(humidity) else humidity
    humidity = mean_filler.get(state+county+city, np.nan) if np.isnan(humidity) else humidity
    humidity = mean_filler.get(state+county, np.nan) if np.isnan(humidity) else humidity
    humidity = mean_filler.get(state, np.nan) if np.isnan(humidity) else humidity
  humidity_update.append(humidity)
humidity_update_nan = len([xi for xi in humidity_update if np.isnan(xi)])
print('nan cases', humidity_update_nan, 'nan percentage', 100 * humidity_update_nan / len(humidity_update))
print()


STARTING CONDITION WITH THRESHOLD OF 5
nan cases 59173 nan percentage 1.989453104643559

nan cases 0 nan percentage 0.0



In [ ]:
us_accident_data['Humidity(%)'] = humidity_update

In [ ]:
# FILL BY WEATHER CONDITION & SHORTER
threshold = 5
pressure_update = []
sample_filler = {}
for pressure, weather, zipcode, city, county, state in zip(us_accident_data['Pressure(in)'], us_accident_data.Weather_Condition, us_accident_data.Zipcode, us_accident_data.City, us_accident_data.County, us_accident_data.State):
  if not np.isnan(pressure) and type(weather) == str:
    # CONSIDERING WEATHER CONDITION
    sample_filler.setdefault(weather+state, [0,0])
    sample_filler.setdefault(weather+state+county, [0,0])
    sample_filler.setdefault(weather+state+county+city, [0,0])
    sample_filler.setdefault(weather+zipcode, [0,0])
    # ADD WINDSPEED
    # CONSIDERING WEATHER CONDITION
    sample_filler[weather+state][0] += pressure
    sample_filler[weather+state+county][0] += pressure
    sample_filler[weather+state+county+city][0] += pressure
    sample_filler[weather+zipcode][0] += pressure
    # COUNT CASES
    # CONSIDERING WEATHER CONDITION
    sample_filler[weather+state][1] += 1
    sample_filler[weather+state+county][1] += 1
    sample_filler[weather+state+county+city][1] += 1
    sample_filler[weather+zipcode][1] += 1
  if not np.isnan(pressure):  
    # CONSIDERING ONLY LOCATION
    sample_filler.setdefault(state, [0,0])
    sample_filler.setdefault(state+county, [0,0])
    sample_filler.setdefault(state+county+city, [0,0])
    sample_filler.setdefault(zipcode, [0,0])
    # ADD WINDSPEED
    # CONSIDERING ONLY LOCATION
    sample_filler[state][0] += pressure
    sample_filler[state+county][0] += pressure
    sample_filler[state+county+city][0] += pressure
    sample_filler[zipcode][0] += pressure
    # COUNT CASES
    # CONSIDERING ONLY LOCATION
    sample_filler[state][1] += 1
    sample_filler[state+county][1] += 1
    sample_filler[state+county+city][1] += 1
    sample_filler[zipcode][1] += 1

# FIND MEAN WHEN THE CASES ARE GREATER THAN THE THRESHOLD
mean_filler = {key:value[0]/value[-1] for key, value in sample_filler.items() if value[-1] >= threshold}


print()
print(f'STARTING CONDITION WITH THRESHOLD OF {threshold}')
print('nan cases', len(us_accident_data['Pressure(in)'][us_accident_data['Pressure(in)'].isna()]), 'nan percentage', 100 * len(us_accident_data['Pressure(in)'][us_accident_data['Pressure(in)'].isna()]) / len(us_accident_data['Pressure(in)']))
print()
for pressure, weather, zipcode, city, county, state in zip(us_accident_data['Pressure(in)'], us_accident_data.Weather_Condition, us_accident_data.Zipcode, us_accident_data.City, us_accident_data.County, us_accident_data.State):
  if np.isnan(pressure):
    if type(weather) == str: # TRY FILLING WITH WEATHER CONDITION
      pressure = mean_filler.get(weather+zipcode, np.nan) if np.isnan(pressure) else pressure
      pressure = mean_filler.get(weather+state+county+city, np.nan) if np.isnan(pressure) else pressure
      pressure = mean_filler.get(weather+state+county, np.nan) if np.isnan(pressure) else pressure
      pressure = mean_filler.get(weather+state, np.nan) if np.isnan(pressure) else pressure
    pressure = mean_filler.get(zipcode, np.nan) if np.isnan(pressure) else pressure
    pressure = mean_filler.get(state+county+city, np.nan) if np.isnan(pressure) else pressure
    pressure = mean_filler.get(state+county, np.nan) if np.isnan(pressure) else pressure
    pressure = mean_filler.get(state, np.nan) if np.isnan(pressure) else pressure
  pressure_update.append(pressure)
pressure_update_nan = len([xi for xi in pressure_update if np.isnan(xi)])
print('nan cases', pressure_update_nan, 'nan percentage', 100 * pressure_update_nan / len(pressure_update))
print()


STARTING CONDITION WITH THRESHOLD OF 5
nan cases 48142 nan percentage 1.6185802876945603

nan cases 0 nan percentage 0.0



In [ ]:
us_accident_data['Pressure(in)'] = pressure_update

In [ ]:
# FILL BY WEATHER CONDITION & SHORTER
threshold = 5
visibility_update = []
sample_filler = {}
for visibility, weather, zipcode, city, county, state in zip(us_accident_data['Visibility(mi)'], us_accident_data.Weather_Condition, us_accident_data.Zipcode, us_accident_data.City, us_accident_data.County, us_accident_data.State):
  if not np.isnan(visibility) and type(weather) == str:
    # CONSIDERING WEATHER CONDITION
    sample_filler.setdefault(weather+state, [0,0])
    sample_filler.setdefault(weather+state+county, [0,0])
    sample_filler.setdefault(weather+state+county+city, [0,0])
    sample_filler.setdefault(weather+zipcode, [0,0])
    # ADD WINDSPEED
    # CONSIDERING WEATHER CONDITION
    sample_filler[weather+state][0] += visibility
    sample_filler[weather+state+county][0] += visibility
    sample_filler[weather+state+county+city][0] += visibility
    sample_filler[weather+zipcode][0] += visibility
    # COUNT CASES
    # CONSIDERING WEATHER CONDITION
    sample_filler[weather+state][1] += 1
    sample_filler[weather+state+county][1] += 1
    sample_filler[weather+state+county+city][1] += 1
    sample_filler[weather+zipcode][1] += 1
  if not np.isnan(visibility):  
    # CONSIDERING ONLY LOCATION
    sample_filler.setdefault(state, [0,0])
    sample_filler.setdefault(state+county, [0,0])
    sample_filler.setdefault(state+county+city, [0,0])
    sample_filler.setdefault(zipcode, [0,0])
    # ADD WINDSPEED
    # CONSIDERING ONLY LOCATION
    sample_filler[state][0] += visibility
    sample_filler[state+county][0] += visibility
    sample_filler[state+county+city][0] += visibility
    sample_filler[zipcode][0] += visibility
    # COUNT CASES
    # CONSIDERING ONLY LOCATION
    sample_filler[state][1] += 1
    sample_filler[state+county][1] += 1
    sample_filler[state+county+city][1] += 1
    sample_filler[zipcode][1] += 1

# FIND MEAN WHEN THE CASES ARE GREATER THAN THE THRESHOLD
mean_filler = {key:value[0]/value[-1] for key, value in sample_filler.items() if value[-1] >= threshold}


print()
print(f'STARTING CONDITION WITH THRESHOLD OF {threshold}')
print('nan cases', len(us_accident_data['Visibility(mi)'][us_accident_data['Visibility(mi)'].isna()]), 'nan percentage', 100 * len(us_accident_data['Visibility(mi)'][us_accident_data['Visibility(mi)'].isna()]) / len(us_accident_data['Visibility(mi)']))
print()
for visibility, weather, zipcode, city, county, state in zip(us_accident_data['Visibility(mi)'], us_accident_data.Weather_Condition, us_accident_data.Zipcode, us_accident_data.City, us_accident_data.County, us_accident_data.State):
  if np.isnan(visibility):
    if type(weather) == str: # TRY FILLING WITH WEATHER CONDITION
      visibility = mean_filler.get(weather+zipcode, np.nan) if np.isnan(visibility) else visibility
      visibility = mean_filler.get(weather+state+county+city, np.nan) if np.isnan(visibility) else visibility
      visibility = mean_filler.get(weather+state+county, np.nan) if np.isnan(visibility) else visibility
      visibility = mean_filler.get(weather+state, np.nan) if np.isnan(visibility) else visibility
    visibility = mean_filler.get(zipcode, np.nan) if np.isnan(visibility) else visibility
    visibility = mean_filler.get(state+county+city, np.nan) if np.isnan(visibility) else visibility
    visibility = mean_filler.get(state+county, np.nan) if np.isnan(visibility) else visibility
    visibility = mean_filler.get(state, np.nan) if np.isnan(visibility) else visibility
  visibility_update.append(visibility)
visibility_update_nan = len([xi for xi in visibility_update if np.isnan(xi)])
print('nan cases', visibility_update_nan, 'nan percentage', 100 * visibility_update_nan / len(visibility_update))
print()


STARTING CONDITION WITH THRESHOLD OF 5
nan cases 65691 nan percentage 2.2085945261713964

nan cases 0 nan percentage 0.0



In [ ]:
us_accident_data['Visibility(mi)'] = visibility_update

In [ ]:
print('NAN %')
us_accident_data.isnull().sum().sort_values()

NAN %


Unnamed: 0                     0
Humidity(%)                    0
Pressure(in)                   0
Visibility(mi)                 0
Wind_Speed(mph)                0
Amenity                        0
Bump                           0
Crossing                       0
Give_Way                       0
Junction                       0
No_Exit                        0
Railway                        0
Roundabout                     0
Station                        0
Stop                           0
Traffic_Calming                0
Traffic_Signal                 0
Turning_Loop                   0
Timezone                       0
Country                        0
Temperature(F)                 0
State                          0
ID                             0
Source                         0
Severity                       0
Start_Time                     0
End_Time                       0
Start_Lat                      0
Start_Lng                      0
Zipcode                        0
Distance(m

In [ ]:
print('UNIQUE VALUES')
us_accident_data.nunique().sort_values()

UNIQUE VALUES


Country                        1
Turning_Loop                   1
Astronomical_Twilight          2
Amenity                        2
Nautical_Twilight              2
Bump                           2
Crossing                       2
Junction                       2
No_Exit                        2
Railway                        2
Roundabout                     2
Give_Way                       2
Stop                           2
Traffic_Calming                2
Traffic_Signal                 2
Civil_Twilight                 2
Sunrise_Sunset                 2
Station                        2
Side                           3
Source                         3
Severity                       4
Timezone                       4
TMC                           21
Wind_Direction                24
State                         49
Weather_Condition            120
Precipitation(in)            256
Wind_Chill(F)                971
County                      1702
Airport_Code                1995
Visibility

In [ ]:
# ELIMINATE NAN'S OF WEATHER CONDITION BECAUSE IT IS VERY HARD TO RECOVER
us_accident_data[us_accident_data.Side == ' '] = np.nan
us_accident_data = us_accident_data.dropna(subset=['Side', 'Weather_Condition', 'Nautical_Twilight', 'Sunrise_Sunset',  'Civil_Twilight', 'Astronomical_Twilight'])
# ELIMINATE NON VALUABLE COLUMNS
# PREDICTORS LIKE DISTANCE
# UNIQUE VALUES AS COUNTRY OR TURNING LOOP
# NON RELATED VARIABLES AS SOURCE, AIRPORT_CODE (THEY MEASSURE THE WEATHER)
# IT IS USELESS TO ESTIMATE THE NANS < 0,5


us_accident_data = us_accident_data.drop('ID', 'Turning_Loop', 'Distance(mi)', 'Description', 'Street', 'End_Time', 'Source', 'Country', 'Airport_Code', 'Weather_Timestamp', 'Wind_Direction', 'TMC', 'Wind_Chill(F)', 'Number', 'End_Lng', 'End_Lat', 'Precipitation(in)'], axis = 1)
us_accident_data.Zipcode = us_accident_data.Zipcode.astype(int)
print('Nan Dataset Quantity')
us_accident_data.isnull().sum().sort_values()

Nan Dataset Quantity


Severity                 0
Civil_Twilight           0
Sunrise_Sunset           0
Traffic_Signal           0
Traffic_Calming          0
Stop                     0
Station                  0
Roundabout               0
Railway                  0
No_Exit                  0
Junction                 0
Give_Way                 0
Crossing                 0
Bump                     0
Amenity                  0
Weather_Condition        0
Wind_Speed(mph)          0
Visibility(mi)           0
Pressure(in)             0
Humidity(%)              0
Temperature(F)           0
Timezone                 0
Zipcode                  0
State                    0
County                   0
City                     0
Side                     0
Start_Lng                0
Start_Lat                0
Start_Time               0
Nautical_Twilight        0
Astronomical_Twilight    0
dtype: int64

In [ ]:
print('UNIQUE VALUES')
us_accident_data.nunique().sort_values()

UNIQUE VALUES


Astronomical_Twilight          2
Bump                           2
Crossing                       2
Give_Way                       2
Junction                       2
No_Exit                        2
Railway                        2
Nautical_Twilight              2
Roundabout                     2
Stop                           2
Traffic_Calming                2
Side                           2
Traffic_Signal                 2
Sunrise_Sunset                 2
Civil_Twilight                 2
Station                        2
Amenity                        2
Severity                       4
Timezone                       4
State                         49
Weather_Condition            120
County                      1687
Visibility(mi)              2231
Pressure(in)                2576
Temperature(F)              4431
Humidity(%)                 4564
Wind_Speed(mph)             6714
City                       11378
Zipcode                    20733
Start_Lng                 965414
Start_Lat 

In [ ]:
# SAVE PRUNED DATASET
us_accident_data.to_csv('/content/drive/My Drive/us_accident_data_pruned.csv')

In [ ]:
# LOAD PROGRESS
us_accident_data = pd.read_csv('/content/drive/My Drive/us_accident_data_pruned.csv')